### PR TITLE
Eng 12741 NP transaction benchmark

### PR DESCRIPTION
**Pull request by vittvolt**
## Intro
- Based on https://github.com/VoltDB/voltdb/pull/3822
- Change the parameters in run.sh when running the benchmark
    - **scale**: Percentage of 2P txns compared to SP txns.
    - **mprate**: Percentage of MP txns compared to SP txns.
    - **skew**: Data skew ratio.
    - **cardcount**: # of cards in the run.

- Usage
    - Make sure the voltdb/bin is in your PATH
    - In one terminal, run `./run.sh prepare` followed by `./run.sh start`
    - In the 2nd terminal, run `./run.sh load` followed by `./run.sh client`

## Pro
- Refer to https://github.com/VoltDB/pro/pull/1969

### CODE REVIEW

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+/**
```
Comment (xinjiacs) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java]: Apache license?

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    public long run(String from_pan,
+                    String to_pan,
+                    double amount,
+                    String currency) throws VoltAbortException {
+
+        long result = 0;
+
+        voltQueueSQL(getAcct, EXPECT_ZERO_OR_ONE_ROW, from_pan);
+        voltQueueSQL(getAcct, EXPECT_ZERO_OR_ONE_ROW, to_pan);
+
```
Comment (xinjiacs) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java]: Do we need run the READs first and abort if the transfer amount is larger than the balance from the from_pan?

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (xinjiacs) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java]: Apache?

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+public class Select extends VoltProcedure {
+
+    public final SQLStmt selAcct = new SQLStmt("SELECT * FROM card_account WHERE pan >= ? and pan < ?;");
+
+    public final SQLStmt selActv = new SQLStmt("SELECT * FROM card_activity WHERE pan >= ? and pan < ?;");
+
+    public long run (String start, String end) {
+        voltQueueSQL(selAcct, EXPECT_NON_EMPTY, start, end);
+        voltQueueSQL(selActv, start, end);
+        VoltTable results[] = voltExecuteSQL(true);
```
Comment (xinjiacs) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java]: What's going on here? Just run two selects and do nothing?

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Redeem.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (xinjiacs) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Redeem.java]: Same here?

File: tests/test_apps/nptxn-perf-benchmark/.gitignore [hunk 6]
```diff
@@ -0,0 +1,5 @@
+*.class
+*.jar
+*.txt
+
+/stats
```
Comment (xinjiacs) [inline tests/test_apps/nptxn-perf-benchmark/.gitignore]: Why .gitignore file here? 
If you run script 'run.sh clean', will these generated files get deleted?

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java (hunk 3 above)
Comment (vittvolt) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java]: @xinjiacs The files from https://github.com/VoltDB/voltdb/pull/3822 had Apache licenses, shouldn't I keep it ?

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java (hunk 2 above)
Comment (vittvolt) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java]: @xinjiacs I set the transfer amount to be very small, so I assume there should be a few amount of aborts. And if we run read first, then we don't have to abort anyway.

Comment (vittvolt) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Transfer.java]: The check is put at the end currently, and to abort the whole transaction if necessary.

File: tests/test_apps/nptxn-perf-benchmark/run.sh [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+if [ $# -gt 1 ]; then
+    for ((i=2; i<=$#; i++)); do
+    	# ${!i} reference the value in the input args
+		cmdargs+=" ${!i}"
+	done
+fi
+
+echo "${0}: Performing ${1} ${cmdargs}..."
+${1} ${cmdargs}
+
```
Comment (vittvolt) [inline tests/test_apps/nptxn-perf-benchmark/run.sh]: issue resolved.

File: tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package np;
+
```
Comment (vittvolt) [inline tests/test_apps/nptxn-perf-benchmark/procs/np/Select.java]: issue resolved

**Comment by xinjiacs** [comment]
@vittvolt Try to add a client thread count to saturate the server. It would benefit us adding scalable 2P transactions in future.

**Pull request merged**